deps/src/serial_relay.c

```diff
@@ -0,0 +1,361 @@
+        debug_print("Read %d bytes: %.*s\n", bytes_read, bytes_read, relay->msg_pub_buffer);
+
+        // Construct zmq message and copy data from buffer
+        zmq_msg_t msg;
+        size_t msg_size = bytes_read;
+        flag = check_zmq(zmq_msg_init_size(&msg, msg_size));
+        if (flag != SR_OK) return flag;
+
+        memcpy(zmq_msg_data(&msg), relay->msg_pub_buffer, msg_size);
+
+        // Relay those bytes through zmq
+        flag = check_zmq(zmq_msg_send(&msg, relay->serial_pub_socket, 0));
+        if (flag != SR_OK) return flag;
+    }
+
+    return SR_OK;
+}
+
+
+enum sr_return relay_read(void *relay)
+{
+    return _relay_read((serial_zmq_relay *)relay);
+}
+
+
+enum sr_return _relay_write(serial_zmq_relay *relay)
+{
+    enum sr_return flag;
+    int bytes_writen;
+
+    // Initialize ZMQ message
+    zmq_msg_t msg;
+    flag = check_zmq(zmq_msg_init(&msg));
+    if (flag != SR_OK) return flag;
+
+    // Check if a message is available to be received from socket
+    int rc = zmq_msg_recv(&msg, relay->serial_sub_socket, ZMQ_DONTWAIT);
+
+    // If we heard a message:
+    if (rc != -1)
+    {
+        // Check size of recieved message to make sure it can be copied into our buffer
+        size_t msg_size = zmq_msg_size(&msg);
+        if (msg_size > PORT_BUFFER_SIZE)
+        {
+            fprintf(stderr, "Message size is too large to write to serial port!");
+            return SR_ERR_MEM;
+        }
+
+        // Copy the msg into the buffer and free ZMQ message
+        memcpy(relay->msg_sub_buffer, zmq_msg_data(&msg), msg_size);
+        flag = check_zmq(zmq_msg_close(&msg));
+        if (flag != SR_OK) return flag;
+
+        // Write message's bytes to ZMQ port
+        bytes_writen = sp_blocking_write(relay->port,
+                                         relay->msg_sub_buffer,
+                                         msg_size,
+                                         1000);
+        flag = check_serial(bytes_writen);
+        if (flag != SR_OK)
+            return flag;
+
+        debug_print("Wrote %d bytes: %.*s\n", bytes_writen, bytes_writen, relay->msg_sub_buffer);
+    }
+
+    return SR_OK;
+}
+
+
+enum sr_return relay_write(void *relay)
+{
+    return _relay_write((serial_zmq_relay *)relay);
+}
+
+
+enum sr_return _close_relay(serial_zmq_relay *relay)
+{
+    // Result check int
+    enum sr_return flag;
+
+    // Free the buffers
+    free(relay->msg_pub_buffer);
+    free(relay->msg_sub_buffer);
+
+    // Close the zmq sockets and context
+    flag = check_zmq(zmq_close(relay->serial_sub_socket));
+    if (flag != SR_OK) return flag;
+
+    flag = check_zmq(zmq_close(relay->serial_pub_socket));
+    if (flag != SR_OK) return flag;
+
+    flag = check_zmq(zmq_ctx_destroy(relay->context));
+    if (flag != SR_OK) return flag;
+
+    // Close out the serial port
+    flag = check_serial(sp_close(relay->port));
+    if (flag != SR_OK) return flag;
+
+    sp_free_port(relay->port);
+
```

Comment:
no `free` on the relay itself? It's usually  best practice to mirror the allocs in the initializer.